### PR TITLE
[IMP] website: improve UX of the WebsiteLoader

### PR DESCRIPTION
**Pull request by robinlej**
When the loader takes a long time before having a website ready to be
shown (due to many features selected), it is confusing for users who
think the page stopped working. They tend to refresh the page, provoking
bugs. To improve the UX, this PR adds:

- a browser warning to prevent the user from closing the tab/refreshing
- several waiting sentences, depending on the features chosen by the
user (the screen switches between them every 10 seconds)
- a loading bar whose progression shows the number of installed modules
- a solid background instead of a transparent one as there is no value
in seeing the preceding background at that point

This PR also removes the secondary loader, once the website has been
created, to avoid the bad UX of having one long loading, then a refresh,
then another loading with the same look.

task-3100223

### CODE REVIEW

**Comment by qsm-odoo** [comment]
@robodoo delegate=bvr-odoo

**Review by bvr-odoo — CHANGES_REQUESTED** [review]
I reviewed the PR, but I don't feel confident enough in my OWL skills to make the final decision on it. That's why I asked @detrouxdev to also take a look before merging it :wink:

File: addons/website/static/src/components/website_loader/website_loader.xml
Comment (robinlej) [reply addons/website/static/src/components/website_loader/website_loader.xml:31]: This is actually the convention for CSS custom properties targetting OWL modules (although I see now that I forgot the second _): https://www.odoo.com/documentation/16.0/contributing/development/coding_guidelines.html#css-variables

File: addons/website/static/src/components/website_loader/website_loader.scss
Comment (robinlej) [reply addons/website/static/src/components/website_loader/website_loader.scss:39]: I lightened the color because the contrast (in terms of accessibility) was not sufficient with the base one.
Maybe I could directly use the dark mode color in there?

File: addons/website/static/src/components/website_loader/website_loader.scss
Comment (robinlej) [reply addons/website/static/src/components/website_loader/website_loader.scss:48]: I don't think it is, considering that both values are computed at configuration start and that valuenow is constantly evolving.

File: addons/website/static/src/components/website_loader/website_loader.scss
Comment (robinlej) [reply addons/website/static/src/components/website_loader/website_loader.scss:42]: Since its part of the specs I did not remove it, but I now apply the default primary color of the dark mode.
In order to avoid a clash of colors between the highlighted part of the title and the tour pointer, I also overrode its color with the dark mode one.

**Comment by qsm-odoo** [comment]
@robodoo delegate=sobo-odoo

**Review by sobo-odoo — CHANGES_REQUESTED** [review]
Here is my review :wink:

File: addons/website/static/src/components/website_loader/website_loader.js
Comment (robinlej) [reply addons/website/static/src/components/website_loader/website_loader.js:34]: Yes, exactly. If you reassign the whole value, it loses track of the changes as it doesn't consider it to be the same object.

File: addons/website/static/src/components/website_loader/website_loader.js
Comment (robinlej) [reply addons/website/static/src/components/website_loader/website_loader.js:122]: It's used so the user doesn't see the spinning wheel when the rpc call is a bit long. (Otherwise, it blocks the page with a spinner until it receives the result.)

File: addons/website/static/src/components/website_loader/website_loader.js
Comment (robinlej) [reply addons/website/static/src/components/website_loader/website_loader.js:125]: Yes, let's say that it is a calculated risk. But in effect there's not much chance of it happening to this extent (last in, first out), it's much more likely to just skip one step every once in a while. And anyway it's mostly a cosmetic / UX issue, so I think the trade is fair.
To remove the `setInterval` and actually await each result, the alternative (I think) would be long polling, which I don't think is actually necessary here.

File: addons/website/static/src/components/website_loader/website_loader.js
Comment (robinlej) [reply addons/website/static/src/components/website_loader/website_loader.js:61]: About the looping: I asked PDE about it already, his answer was "nevermind". It's not absolutely ideal, but not a big deal either.
About trying to match the ratio of installed features with the tip displayed: since it's not a smooth curve and some features / dependencies took way less time than others, this could switch too quickly, or skip some tips, and generally speaking in terms of UX I think it's better to change on a regular basis (at least more or less) than have widely different timings. Especially because the first modules are the ones which take more time, and the idea is to tell the user "be patient, it's working", so we need to convey this message soon enough. By the time it loops, the progress bar should be advanced enough that the tips don't really matter anymore.

File: addons/website/controllers/backend.py
Comment (rdeodoo) [inline addons/website/controllers/backend.py:67]: This part

File: addons/website/controllers/backend.py
Comment (robinlej) [reply addons/website/controllers/backend.py:55]: Answering a few points (not all) before I jump back into the commit and check the rest:
- The dark mode is actually only available on enterprise, so that purple is the enterprise one.
- As I said in another comment, the tips are more a sign of "please be patient" than anything else. So not a big deal if they don't all appear. :)
- To test on local, you should launch the server with the `--workers` flag (I put 4, I guess any number could work). Otherwise, it cannot install the modules and track their installation in parallel. ;)

File: addons/website/controllers/backend.py
Comment (robinlej) [reply addons/website/controllers/backend.py:68]: I might as well rename it, there's really no need to shorten and obfuscate it. :shrug:

**Review by rdeodoo — CHANGES_REQUESTED** [review]
I was just passing by due to mail notification and the fact that this task is very interesting.
I am the one that suggested the loading bar spec so I had a look and I don't think the current implementation really makes sense as it's only considering the selected apps, while it should be looking for all apps.

If you take a look at this video: https://watch.screencastify.com/v/jCsGPVxnsqt26Q31RQ4I:
- Configurator loading start at 00:22
- The loading bar remains empty until 1:41
- At 1:41 it fill up to full instantly
- It actually finished at 1:53

But if you look your server logs, there will be many many things done, especially those kind of things:
```
2023-06-28 09:43:02,080 20095 INFO masterLoaderRole odoo.modules.loading: Loading module sales_team (28/58) 
2023-06-28 09:43:15,113 20095 INFO masterLoaderRole odoo.modules.loading: Loading module account_payment (43/58)
```

This is those numbers I wanted to use to generate the loading bar, for instance 28/58 = 48%.

I am not sure if there is an easy way (or a way at all) to find those numbers which are shown in the logs, it would need to be investigated.

In fine, what is important is the flow I shown in the video, which is our main flow and the one we try to improve: the user has only website installed and select some apps and features in the configurator which will install some (many?) apps which takes a lot of time.
If you install ecommerce, it will install accounting etc which are very big modules.
We need to make the bar responsive and evolving, not going from 0% to 100% in a blink at the end of the configurator loading time, else it does not bring any value and could be removed from the task/spec to only keep the rotating wording.

Note: it sometimes jumps from (for instance) 43/58 to 72/120, I don't think it really matters much -> worst case the bar is reducing a bit, best case you never reduce but only wait for the % to be higher to adapt the progress bar.

File: addons/website/static/src/components/website_loader/website_loader.js
Comment (robinlej) [reply addons/website/static/src/components/website_loader/website_loader.js:203]: - `selectedFeatures` is only supposed to change once, but the method was indeed messy at best. I'll extract the tips and the refactored function outside of it. But you'll see I opted for a plain array for 2 reasons: 1) it's easier to manipulate and 2) to avoid making useless manipulations later on, the 2 non-feature related tips are also in there.
- I thought about using `t-out` to inject the name inside the title, etc. directly from the XML, but I would need to be able to replace `%s` from the XML to do so and I don't think I can? (I need the `%s` for translations, where the word order might differ.)

File: addons/website/static/src/components/website_loader/website_loader.xml
Comment (robinlej) [reply addons/website/static/src/components/website_loader/website_loader.xml:8]: I think it's a remain of an error I got, but it doesn't seem necessary anymore.

File: addons/website/controllers/backend.py
Comment (robinlej) [reply addons/website/controllers/backend.py:55]: > And when the real one then starts (after having configured everything) both "interval" seem to exist at the same time (even if the tips displayed seem correct) open_mouth

Do you mean both `tipsInterval`s? The first one is cleared when the `useEffect` is called again, with `return () => clearInterval(tipsInterval);`.

File: addons/website/static/src/components/website_loader/website_loader.js
Comment (sobo-odoo) [reply addons/website/static/src/components/website_loader/website_loader.js:39]: Okay :+1:  (I am not very good at asynchronization haha)

File: addons/website/controllers/backend.py
Comment (sobo-odoo) [reply addons/website/controllers/backend.py:55]: > The dark mode is actually only available on enterprise

Ah yes that's right haha
Anyway if PDE is okay with this color, then it's okay for me too :wink: 
 
 > To test on local, you should launch the server with the --workers flag
 
 Thanks for the tip :smile: 
 
 > Do you mean both tipsIntervals? The first one is cleared when the useEffect is called again, with return () => clearInterval(tipsInterval);.
 
 My bad, I don't know how I saw that but the issue I mentioned does in fact not happen :no_mouth:

File: addons/website/static/src/components/website_loader/website_loader.js
Comment (robinlej) [reply addons/website/static/src/components/website_loader/website_loader.js]: I've reworked the flow, so this is done now :+1:

File: addons/website/controllers/backend.py
Comment (robinlej) [reply addons/website/controllers/backend.py:55]: > On the first page of the configurator, so the loading hasn't even started yet, the useEffect applied on selectedFeatures is triggered and so the waitingTips interval function is triggered too, browsing the tips array for "nothing"

With the addition of a `showWaitingMessages` state/prop, this should be good now and only apply the effects if it is specifically requested with the `showLoader()` function.

File: addons/website/static/src/components/website_loader/website_loader.js
Comment (sobo-odoo) [reply addons/website/static/src/components/website_loader/website_loader.js]: - As said in my "behavior" comment, it was called at least twice (as we discussed it :wink:)
- You are right, I didn't think about this translation issue, it can stay like that then :+1:

File: addons/website/controllers/backend.py
Comment (robinlej) [reply addons/website/controllers/backend.py:67]: As discussed:
1. For the progressbar to work, you should launch the local server with `--workers 4`.
2. I haven't found a way to directly get those numbers shown in the logs. They are generated on the fly in `odoo/modules/loading.py` (search for "Loading module"), but there doesn't seem to be an easy way to hook up to these from outside. (But if I'm wrong I'll be glad to know!)

Comment (rdeodoo) [reply addons/website/controllers/backend.py:67]: I don't see how the progressbar could work at all with this implementation.

Let's take the main flow: installing only one app, and probably ecommerce that is.

The current code will be looking for when the `website_sale` module is installed before updating the progressbar.
There is conceptually 2 problems:
1. It means that there is only 2 states in the progress bar: 0% when website_sale is not installed and 100% when it is installed. This means that it will not be a progress bar at all, the goal was (and is) to show a.. **progress** bar, meaning it should show you how is the progress going: 1% 4% 14% 22% 30% etc etc, it should be moving during all the processing time.
The goal of this task is to avoid the user thinking something is stuck and that the page is not working anymore. This will be doing the opposite as this new element on the UI will also look stuck now. 
2. On top of having a limited steps in the progress bar (basically the number of selected apps), all those steps will be reached only after ~90% of the process, because the majority of the processing time is installing dependency apps which are not the selected apps.
For instance, installe `website_sale` will install many many many many apps like accounting and so on, which won't have an impact on the progress bar.

If we can't find a solution, this progress bar can simply be removed from the PR IMHO, it's just something I pushed for on the spec and that was not supposed to be there in the first place, so if it's not really helpful and adding value, we might as well get rid of it (less code).

I will have a look to see if I can find a way to make it somehow work like that.

Comment (rdeodoo) [reply addons/website/controllers/backend.py:76]: I actually pushed this commit https://github.com/odoo/odoo/pull/118525/commits/2ede242258d27034d5e86eafd5fd78788bbff7be
(Just a crappy wip)

Which basically do:
```py
        a = request.env['website.configurator.feature'].browse(selected_features).module_id.upstream_dependencies(exclude_states=('',))
        features_info = {
            'total': len(a),
            'nbInstalled': len(a.filtered(lambda f: f.state == 'installed')),
        }
        return features_info
```
Which seems to almost behaves as expected:
- It's missing the auto install modules which are not depedencies of any module like web_unsplash, so I think there is a discrepancy between the logs and what we got here (like ~35 vs ~55) because of that -> Maybe to check if can be fixed? Otherwise maybe not the end of the world 
- I think the % will start at like 50% because a good junk of the apps are already installed (base, website, portal etc).
Need to investigate if we can have the list of installed app before applying the configurator, so we know which one to remove from the computation.

Note that I am surprised we have access to those installed module in real time (I mean in a rpc while the module install process is still pending) -> I would have expect this install to be inside a transaction and so nothing commited before the end and so any other cursor not aware of it, but I guess some stuff are commited during the install.

File: addons/website/static/src/components/website_loader/website_loader.js
Comment (rdeodoo) [inline addons/website/static/src/components/website_loader/website_loader.js:48]: 2000 was a bit high imo, 300 is too small and was just for the test/wip.
I would expect the configurator to take like ~4-10 sec max if you select 2-4 apps on the saas /trial, could you have a look?
But maybe 2000 was fine, if the configurator is fast, the user probably won't leave anyway, and if it's slow, 2 sec for an update is okay. I guess you discussed that with @sobo-odoo and 2 sec was judged fine anyway? You can probably ignore this :+1:

File: addons/website/static/src/components/website_loader/website_loader.xml
Comment (rdeodoo) [inline addons/website/static/src/components/website_loader/website_loader.xml:30]: Was for debugging purpose, but wouldn't it be nice to show to apps total (instead of %) ?
Like [23/45], then [33/45] etc?
I think I wrote something about it in the spec, it would give the impression than something is being done (and would be more "understandable" than % -> You know there is 45 moduels to be installed, and this is doing the install of the 23th one

**Review by rdeodoo — COMMENTED** [review]
I pushed a commit that might help you having a smoother and working progress bar, you can probably investigate it.
I left some comments too

File: addons/website/static/src/components/website_loader/website_loader.xml
Comment (robinlej) [reply addons/website/static/src/components/website_loader/website_loader.xml:30]: That's +/- what I initially did, but @pierre-pde didn't like it. He considers that the progressbar gives enough information in itself.

I'll look at the rest on Monday, thanks!

Comment (rdeodoo) [reply addons/website/static/src/components/website_loader/website_loader.xml:30]: If we have the progress bar working smoothly (and not going from 0 to 100 in one go at 90% of the process) it's maybe something we can live with indeed.
Even if I think having the number increase every 2 sec will be noticeable, while the progress bar moving 3% every 2 seconds is barely noticeable in itself, but as you prefer.

**Review by sobo-odoo — CHANGES_REQUESTED** [review]
Here is my re-review :wink: 
Since you need to modify the PR to track the installed modules, I am already posting it so you can make the changes at the same time 
It seems to work correctly but I will wait for it to be complete to review and test it fully one last time :wink: (also because the code on github is barely readable now with all the comments in it haha)

File: addons/website/static/src/components/website_loader/website_loader.js
Comment (robinlej) [reply addons/website/static/src/components/website_loader/website_loader.js]: Alright, makes sense. I switched for a timeout within the `trackModules()` function, so that the next is only called when the previous one is done.

File: addons/website/static/src/components/website_loader/website_loader.js
Comment (robinlej) [reply addons/website/static/src/components/website_loader/website_loader.js]: 2000 was more of a default suggestion, I didn't discuss it, but yes what you wrote was my train of thought (it's useful when it's slow, and if it's slow 2s is an acceptable interval to show there is some progress). I've updated it to 1000, it might be a good balance. What do you think?

Btw, I tested the configurator on trial, and it's taking at least as long as a local server or the runbot.

File: addons/website/static/src/components/website_loader/website_loader.xml
Comment (robinlej) [reply addons/website/static/src/components/website_loader/website_loader.xml]: I tested it in various situations, among which the single eCommerce module installation, and you do see the progress.

If I remember correctly, @pierre-pde's opinion was 1) that the modules' numbers don't evoke anything to the base user, 2) that a % is redundant with the progressbar, 3) it takes space for no real added value. So the progressbar is enough.

File: addons/website/static/src/components/website_loader/website_loader.js
Comment (robinlej) [reply addons/website/static/src/components/website_loader/website_loader.js]: Yes, it was a leftover from the previous manipulations I did on it. Thanks.

File: addons/website/static/src/components/website_loader/website_loader.js
Comment (rdeodoo) [reply addons/website/static/src/components/website_loader/website_loader.js]: I don't have a strong opinion, it's a balance between not taking too many ressources and having a smooth progress.
I think anything between 1s and 2s might be correct.
The more I think about it the more I think it was fine (2s) :+1:
I'll let @sobo-odoo decide anyway

File: addons/website/static/src/components/website_loader/website_loader.xml
Comment (robinlej) [reply addons/website/static/src/components/website_loader/website_loader.xml]: I've seen both, but if you tell me that's the official convention, I'll stick to that. :)

File: addons/website/static/src/components/website_loader/website_loader.xml
Comment (rdeodoo) [reply addons/website/static/src/components/website_loader/website_loader.xml]: > I tested it in various situations, among which the single eCommerce module installation, and you do see the progress.

Weird, because you basically had only the direct dependencies of the module in your controller.
So for `website_sale`, you only has like 5-7 modules like UOM etc.
This was because of the `auto_install_required` part of your condition.
So I don't see how it was really behaving smoothly, what am I missing? (and I was in multi workers)

Basically, this is what I experienced with your diff before I pushed my wip: https://watch.screencastify.com/v/jCsGPVxnsqt26Q31RQ4I

> 1) that the modules' numbers don't evoke anything to the base user, 2) that a % is redundant with the progressbar, 3) it takes space for no real added value. So the progressbar is enough.

Well if the progress bar is progressing constantly / smoothly, I don't mind anyway, it will be enough indeed, but just to reply (ignore it):
1.  I am sure everyone understand what X/Y means (especially with a word next to it), it's literally what you see everywhere, PlayStation updates, TV update, phone update etc.
2. Yes but only if the progress bar move constantly
3. Not really, as the progress bar is taking a whole "line" on the screen while it occupied a very minor part of it (like 5-10%?), so adding a small % next to it would not change much. Better, it would be part of the progress bar, inside it.

Comment (robinlej) [reply addons/website/static/src/components/website_loader/website_loader.xml]: About the 1st part: yes, sorry I actually meant _now_ you see the progress with the ecommerce module. I just wanted to say that the case you pointed out was resolved and that I feel the progress on the progress bar itself is noticeable with those adjustments.
(Basically answering to:)
> the progress bar moving 3% every 2 seconds is barely noticeable in itself

Comment (rdeodoo) [reply addons/website/static/src/components/website_loader/website_loader.xml]: It indeed now really looks great :+1: 
See https://watch.screencastify.com/v/GSOqa5LQttsNeds8rIu6 on your latest runbot.

I am a bit skeptical about the part where all modules are installed but there is still some time before the configurator is really finished, a few seconds, but it will probably be okay as 95% of the processing time will be covered by the progress bar, the impression of "this is stuck" should be gone, **let's not care about that for now**, that would be premature optimization. (Could have tried to do something if we kept the % next to the progress bar, at 100% it could have mentioned something like "Module install done, now cleaning xyz", like installation process usually do when you are at 100% but still something being processed.).

File: addons/website/static/src/components/website_loader/website_loader.xml
Comment (sobo-odoo) [reply addons/website/static/src/components/website_loader/website_loader.xml]: I don't know if it is a convention, but the rare examples in the documentation do it like this (a bit under [here](https://www.odoo.com/documentation/master/developer/reference/backend/views.html#attributes) for example ) 

But even if not a convention, you can agree that is easier to read like this, no ? :stuck_out_tongue_closed_eyes:

**Review by sobo-odoo — CHANGES_REQUESTED** [review]
It seems to work well :+1: 
Here is a last review and after that it will be ready to go :wink: (Don't forget to also rebase, as the last change was last week)

File: addons/website/controllers/backend.py
Comment (robinlej) [reply addons/website/controllers/backend.py:55]: - As far as I'm concerned, it's normal: if it's already installed, it shouldn't have any progress to show. The "odd" case is actually having the messages appear once again for every feature that was already installed on the first website when you select new features for the 2nd website − but this case has been deemed OK. :)
- I'll ask PDE for a new check on this. :)

File: addons/website/static/src/components/website_loader/website_loader.js
Comment (robinlej) [reply addons/website/static/src/components/website_loader/website_loader.js:24]: Probably not indeed.

File: addons/website/static/src/components/website_loader/website_loader.js
Comment (robinlej) [reply addons/website/static/src/components/website_loader/website_loader.js:103]: No specific reason to do one over the other. :shrug: I'll adapt them.

File: addons/website/static/src/components/website_loader/website_loader.js
Comment (robinlej) [reply addons/website/static/src/components/website_loader/website_loader.js:144]: It's a class field, not a variable, so just like the functions are declared as `fctName() {}` and not `function fctName() {}`, this doesn't need `const` (see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes#field_declarations).
I guess I could extract it as a constant outside of the component instead. Do you think there should be a reason to do so?

Comment (sobo-odoo) [reply addons/website/static/src/components/website_loader/website_loader.js:144]: Okay indeed :open_mouth:
> Do you think there should be a reason to do so

Not really :thinking: Since it works, it can stay like this I guess :wink:

File: addons/website/controllers/backend.py
Comment (robinlej) [reply addons/website/controllers/backend.py:55]: Final word: I clear the interval when I reach the last index, so that it stays on the last message.

**Review by sobo-odoo — CHANGES_REQUESTED** [review]

**Comment by sobo-odoo** [comment]
@odoo/rd-security could you check this PR please ? :)

**Comment by robinlej** [comment]
@xmo-odoo 

> odoo/addons/website/static/src/client_actions/configurator/configurator.js:380
> window.location.replace(/web#action=website.website_preview&website_id=${encodeURIComponent(resp.website_id)}&enable_editor=1);
> Line may contain security risk: location.replace. Please contact security team for advanced review (you can ping @odoo/rd-security)

The security flag here doesn't concern the part of the line that was modified. I didn't touch the `encodeURIComponent()` part, I only removed a query parameter.

> odoo/addons/website/static/src/components/website_loader/website_loader.js:168
> return markup(sprintf(title, highlight));
> Line may contain security risk: markup(. Please contact security team for advanced review (you can ping @odoo/rd-security)

The variables given here (title & highlight) are hardcoded translated strings (given through an array of objects encoded as a class field), so my understanding is that it's ok.

**Comment by sobo-odoo** [comment]
@xmo-odoo Could you re-check ? :slightly_smiling_face: 
The diff is [here](https://github.com/odoo/odoo/compare/13afea8cfa54574a11a7cd3df0bbe15f4ff64c9b..0915effa5c54195d6d867e279df5d136a795b3bc) for the `markup` flag 
For the `location.replace` flag, in a review of [another PR](https://github.com/odoo/odoo/pull/117664#issuecomment-1498485601), it was considered OK to touch this line because this is a local URL

**Comment by sobo-odoo** [comment]
@xmo-odoo (repinging you in case you didn't see :) )

**Review by sobo-odoo — APPROVED** [review]
Thanks ! 
Let's r+ :partying_face: 

@robodoo r+

**Pull request closed without merging**
